Sources/COWMacros/COWExpansionFactory.swift

```diff
@@ -218,30 +223,196 @@ internal class COWExpansionFactory<Context: MacroExpansionContext> {
+  // the derived storage must also conform to Equatable (otherwise we hit a
+  // compiler bug that breaks the build).
+  // https://forums.swift.org/t/macro-multiple-matching-functions-xyz-error-though-the-expanded-code-has-no-duplicate-functions-and-compiles/68700/4
+  // FIXME: remove the workaround when the compiler bug is fixed.
+  private func applyEquatableWorkaroundForDerivedStorage(
+    members: inout [MemberBlockItemSyntax],
+    protocols: inout [InheritedTypeSyntax]
+  ) {
+    // Bail out if derived storage doesn't conform to Equatable.
+    guard protocols.contains(where: {
+      guard let name = $0.type.identifier else {
+        return false
+      }
+      return equatableProtocolNames.contains(name) ||
+             hashableProtocolNames.contains(name)
+    }) else {
+      return
     }
     
+    // Bail out if `appliedStructDecl` does not manually implement `==`.
+    guard let nestingEqualFunc = appliedStructDecl.memberBlock.members
+      .lazy
+      .map(\.decl)
+      .compactMap({$0.as(FunctionDeclSyntax.self)})
+      .first(where: {$0.likelyToConformToEquatable(for: appliedStructDecl)}) else {
+      return
+    }
+    
+    // Another compiler bug another workaround.
+    // The compiler doesn't call the manually-implemented `==` in `appliedStructDecl`;
+    // it will synthesize one for the derived storage and call it instead.
+    // Expand an `==` that is an exact copy (except for the signature) of
+    // `appliedStructDecl` to make sure the program behavior is correct.
+    
+    // In order to make this workaround work with computed properties and
+    // immutable functions with return values, we also have to copy them into
+    // the derived storage.
+    appliedStructDecl.memberBlock.members
+      .lazy
+      .map(\.decl)
+      .compactMap { $0.as(VariableDeclSyntax.self) }
+      .filter { !$0.isStatic && !$0.isStored }
+      .forEach { members.append(MemberBlockItemSyntax(decl: $0)) }
+    appliedStructDecl.memberBlock.members
+      .lazy
+      .map(\.decl)
+      .compactMap { $0.as(FunctionDeclSyntax.self) }
+      .filter { !$0.isStatic && !$0.isMutating && !$0.returnTypeEquals(to: "Void") }
+      .forEach { members.append(MemberBlockItemSyntax(decl: $0)) }
+    
+    guard let body = nestingEqualFunc.as(FunctionDeclSyntax.self)!.body else {
+      // Could this ever happen (i.e. a function declaration without
+      // implementation in struct is invalid)?
+      fatalError()
+    }
+    // Use signature from proxy protocol if needed. See the comments below
+    // (why we are adding a proxy protocol) for the reason.
+    #if swift(<5.9.2)
+    let nestedEqualFuncSignature: SyntaxNodeString = "static func equalsWorkaround(lhs: Self, rhs: Self) -> Bool"
+    #else
+    let nestedEqualFuncSignature: SyntaxNodeString = "static func == (lhs: Self, rhs: Self) -> Bool"
+    #endif
+    let nestedEqualFunc = try! FunctionDeclSyntax(nestedEqualFuncSignature) {
```

Comment:
Creating syntax from API instead of string literals has a better performance. I recommend to add a `createEqual` or `createEqualForEquatableWorkaround` to encapsulate this process.